### ISSUE TITLE
Research the basic principles of uploading and downloading files

### ISSUE THREAD

**Issue by Oleksii-Kshenskyi** [Oleksii-Kshenskyi/dark_hand#4]
This is an issue for comprehensive research of downloading and uploading files. Most likely it's going to be HTTP upload/download, but at this point I'm not committing to anything yet.

**Research the following:**
- How are files uploaded and downloaded?
- What libraries/packages does Elixir use for uploading/downloading files?
- What external server/service can be used for uploading files to it and downloading from it? The service should ideally be cheap and allow for simple file storage that allows for programmatically downloading to it / uploading from it, without any bloat or closed API of the modern cloud services.

**NOTE:** This issue is extremely important. It is a time-boxed research where no more than [16.8 U @ 07:00:00] should be spent on research itself and no more than [7.2 U @ 03:00:00] on the subsequent planning of the project's architecture, creating the issues and estimations. Only once this research is done, the remaining project can be properly estimated and implementation can be started.

**05/04/2020:**
- Unfortunately, there is no easy way to store files in the modern infrastructure. Each file storage service requires knowledge of their own web API, enforces their own rules on authentication etc.
- Due to this, the most logical thing for me to do would be to pick Microsoft's OneDrive. The reason is that all of my personal files are there already, and therefore integrating Dark Hand with that specific service would make the most sense at this particular moment.
- In general, so far the most interesting functionality I'm hoping to implement for now in the scope of Run One is:
=> Downloading a file from internet (simple direct download of a file by URL);
=> Uploading a file to OneDrive;
=> Downloading a file from OneDrive.
=> Command line interface for all this.
=> If time allows, think about ls-ing a directory on OneDrive and getting metadata of a file.

**05/05/2020:**
- There is two main options in Elixir for handling HTTP requests: httpc and HTTPoison. The difference is that httpc is a part of the standard library (an Erlang lib) while HTTPoison is an external package. Therefore using httpc may be preferable.
- There is an option with httpc to stream your download directly to a file. While HTTPoison seems to first put the whole file into memory and afterwards you have to dump it into a file manually. This kind of behavior may be dangerous when you're downloading extremely large files.
- Unfortunately, after a deeper research, it was decided to use HTTPoison instead. The reasoning is that httpc is an Erlang library, and the docs are only available in Erlang terminology. Using it properly would require to learn part of Erlang and understand Erlang documentation, which is way too time consuming for a month long project.
- Instead of httpc, use HTTPoison and to overcome the "storing whole file in the memory" problem, use async downloads and Elixir Streams. This [article](https://www.poeticoding.com/elixir-streams-to-process-large-http-responses-on-the-fly/) shows how such functionality should be done.

**Conclusions:**
- The file storage service of choice is OneDrive;
- The planned functionality is:
  - Downloading files directly with an HTTP get request (unrelated to OneDrive);
  - Downloading files from OneDrive;
  - Uploading files to OneDrive;
  - ls-ing a directory on OneDrive;
  - Getting stats of a particular file on OneDrive;
  - Some kind of a progress bar for uploads and downloads.
- To enable working on the above OneDrive functionality, first I'm going to need to authenticate on OneDrive. So OneDrive authentication should be the very first thing you implement before doing anything else.
- **NOTE:** out of all of those, only downloading files + OneDrive download + OneDrive upload are critical and mandatory for the run to succeed. Everything else is additional functionality that would be nice to have but is OK if it doesn't make it into the run. The reason is that Run One (as with all the runs, refer to [2020 Challenge](https://github.com/Oleksii-Kshenskyi/2020_challenge/) for more info) is just one month long, and having to learn Elixir + functional programming + web/HTTP is already big enough of a task.

**Comment by Oleksii-Kshenskyi** [comment]
[WEIGHT: 3.181 U @ 01:19:31]

**Comment by Oleksii-Kshenskyi** [comment]
Research done.
